orttraining/orttraining/core/optimizer/memory_alleviation.h

```diff
@@ -0,0 +1,285 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <charconv>
+#include "core/common/inlined_containers.h"
+#include "core/common/string_utils.h"
+#include "core/optimizer/graph_transformer.h"
+
+namespace onnxruntime {
+
+/**
+@Class MemoryAlleviation
+
+Find recomputable subgraphs and enable according to user configs.
+*/
+
+class MemoryAlleviation : public GraphTransformer {
+ private:
+  using NodeOutputPort = std::pair<const Node*, int>;
+  using ActivationUsedMap = InlinedHashMap<std::string, std::pair<bool, bool>>;
+
+  /**
+   * @brief Level to control allowed operations during subgraph detecting.
+   * Level 0: only allow cheap-to-compute operations.
+   * Level 1: allow more expensive operations.
+   */
+  enum class ProbeLevel {
+    Basic = 0,
+    Advanced = 1,
+  };
+
+  /**
+   * @brief Type of memory reduction techniques.
+   */
+  enum class AlleviationType {
```

Comment:
Do we need this given we only have recompute? You might argue that tomorrow we could have another technique, but this code is pretty specific to recompute... so not sure if this is easily extendible to another technique

Reply:
Yeah, good questions! As we synced offline, I cleaned up the logic, by intention, move all recompute related logic tother, I feel about half of the code is recompute specific, the others can be shared to offload later if we need it.